src/SFA.DAS.ProviderApprenticeshipsService.Web/Orchestrators/BulkUploader.cs

```diff
@@ -0,0 +1,165 @@
+            if (!dateMatch.Success || !dateParseSuccess)
+                errors.Add(new UploadError(ApprenticeshipFileValidationText.FilenameFormat.Text, ApprenticeshipFileValidationText.FilenameFormat.ErrorCode));
+
+            else if(outDateTime > DateTime.Now)
+                errors.Add(new UploadError(ApprenticeshipFileValidationText.FilenameFormatDate.Text, ApprenticeshipFileValidationText.FilenameFormatDate.ErrorCode));
+
+            else if (!attachment.FileName.EndsWith(fileEnding) || !attachment.FileName.StartsWith(fileStart))
+                errors.Add(new UploadError(ApprenticeshipFileValidationText.FilenameFormat.Text, ApprenticeshipFileValidationText.FilenameFormat.ErrorCode));
+            
+            if (attachment.ContentLength > maxFileSize)
+                errors.Add(new UploadError(ApprenticeshipFileValidationText.MaxFileSizeMessage(maxFileSize)));
+
+            return errors;
+        }
+
+        public IEnumerable<ApprenticeshipUploadModel> CreateViewModels(HttpPostedFileBase attachment)
+        {
+            string result = new StreamReader(attachment.InputStream).ReadToEnd();
+            using (TextReader tr = new StringReader(result))
+            {
+                var csvReader = new CsvReader(tr);
+                csvReader.Configuration.HasHeaderRecord = true;
+                csvReader.Configuration.IsHeaderCaseSensitive = false;
+
+                try
+                {
+                    return csvReader.GetRecords<CsvRecord>()
+                        .ToList()
+                        .Select(MapTo);
+                }
+                catch (CsvMissingFieldException exception)
+                {
+                    var exceptionData = exception.Data["CsvHelper"];
+                    _logger.Error(
+                        exception,
+                        $"Failed to create files from bulk upload. {typeof(CsvMissingFieldException)} Data CsvHelper {exceptionData}");
+                    throw new Exception("Cannot read all file");
+                }
+                catch (Exception exception)
+                {
+                    var exceptionData = exception.Data["CsvHelper"];
+                    _logger.Error(
+                        exception,
+                        $"Failed to create files from bulk upload. Exception Data CsvHelper {exceptionData}");
+                    throw new Exception("Failed to create apprentices from file");
+                }
+            }
+        }
+
+        public virtual IEnumerable<UploadError> ValidateFields(IEnumerable<ApprenticeshipUploadModel> records, List<ITrainingProgramme> trainingProgrammes)
+        {
+            var errors = new List<UploadError>();
+
+            if (!records.Any()) return new[] { new UploadError(ApprenticeshipFileValidationText.NoRecords) };
+
+            var index = 1;
+            foreach (var record in records)
```

Comment:
Not for this sprint but we should look at the performance of the code with large files. It's very likely we could parallelise some of this.